presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java

```diff
@@ -11,25 +11,22 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.facebook.presto.sql.planner.plan;
+package com.facebook.presto.spi.plan;
 
-import com.facebook.presto.spi.plan.PlanNode;
-import com.facebook.presto.spi.plan.PlanNodeId;
 import com.facebook.presto.spi.relation.VariableReferenceExpression;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.Iterables;
 
 import javax.annotation.concurrent.Immutable;
 
+import java.util.Collections;
 import java.util.List;
 
 import static java.util.Objects.requireNonNull;
 
 @Immutable
-public class ProjectNode
-        extends InternalPlanNode
+public final class ProjectNode
+        extends PlanNode
 {
     private final PlanNode source;
     private final Assignments assignments;
```

Comment:
See the TODO below this comment on line 34 (idk why github wont let me comment on lines that are not part of the diff).

Should that TODO be addressed before we move ProjectNode into SPI?

Reply:
Right now, it might be a little hard. Ideally, we should not put identity assignment in map if an output variable is assigned to an identical variable (i.e. `a = a`). However, we currently have two representations of variable now before the whole migration completes: `SymbolReference` and `VariableReference`, which will makes deciding which form to return very difficult.

Personally I don't feel strongly we should inline the `Assignment` into `ProjectNode` like `TableScanNode`. It is slightly cleaner but you will need to refactor a lot of code depends on it.